### ISSUE TITLE
How to implement PCME + DAA?

### ISSUE THREAD

**Issue by SanghyukChun** [VL-Group/2022-NeurIPS-DAA#2]
Hello, thanks for your excellent work!
I wonder how PCME + DAA in the paper is implemented.

I presume that PCME + DAA replaces the MC soft contrastive loss of PCME to the DAA loss,
while the `sims` is computed by the sampled L2-distance (in this paper, set J=5)

```python
# it is done by Monte-Carlo sampled L2-distance proposed in PCME
sims = self.forward_sim(img_embs, cap_embs, cap_lens)

# The remaining are the same as the current implementation
sems = cider_compute((tfidf_GT, tfidf_single))
self.optimizer.zero_grad()
loss = self.forward_loss(sims, sems)
```

Am I correct? or is there any more implementation detail?

Thanks

**Comment by SanghyukChun** [comment]
Oh, maybe out of the topic in the original issue, but I also wonder how DAA "approximates" probabilistic embeddings for deterministic embeddings. Could you give me a brief idea of how DAA can approximate stochastic embeddings? I have tried to figure out why in the paper, but I couldn't get how DAA and probabilistic embeddings are related. As I understood, DAA and probabilistic embeddings have no correlation, but DAA is another type of loss function based on "Average Semantic Precision (ASP)"

**Comment by leolee99** [comment]
Thank you for your interest in our work. The MC soft contrastive loss is not replaced by DAA loss. We keep the original MC soft contrastive loss as a part of the final loss. Actually, final loss = i2t soft contrastive loss + t2i soft contrastive loss + uniform loss + vib_loss + i2t_cider loss +t2i_cider loss. You just need to add cider loss to the original loss in PCME. Other code is correct.

The reason why DAA approximates probabilistic embeddings, I think as follows. To represent diversity well for probabilistic embedding, the ideal training datasets should be well-annotated. Concretely, there are so many images that are wrongly related to some captions. eg., there are several images about ''some people are playing football'' in MS-COCO, but the captions belonging to image B will be regarded as the negative of image A, even if they semantically match. These wrong relationships will prevent the probabilistic model from fitting real correspondence distribution, leading to damaged diversity.  DAA is a method to excavate extra correspondences out of original annotations. Thus, this can help the probabilistic model training with more annotations and fitting real correspondence distribution better. Afterward, the diversity of probabilistic embedding will be enhanced.
